### PR TITLE
111 file info endpoint

### LINKED ISSUE #111: Get DV File Info  (API endpoint #8)
This endpoint should find and return a `DataverseFileInfo` object . If a `DataverseFileInfo` object doesn’t exist, one is created.

- **Input**: 
			- `DataverseHandoff` object_id
			- `OpenDPUser` object_id 
- **Output**:
	- `DataverseFileInfo` fields in JSON format except `data_profile` and `source_file` 
	   - Include related depositor and analysis information to help determine workflow status
	- Differentiate between creation and update. `is_created`

### Pre-work for DataverseFileInfo:
- Add JSONField `dataset_schema_info` . Does not need encryption. Default for both is blank (or equivalent)
-  Add JSONField `file_schema_info` . Does not need encryption. Default for both is blank (or equivalent)
- Rename `DataverseFileInfo.installation_name` to `DataverseFileInfo.dv_installation`
	- Make it a ForeignKey to a `RegisteredDataverse`
-  Add a django (uniqueconstraint)[https://docs.djangoproject.com/en/3.1/ref/models/constraints/#uniqueconstraint] to `DataverseFileInfo` using the fields: (`dataverse_file_id', 'dv_installation'`)
  
### Basic workflow is:
1. Check the database using `dataverse_file_id` and `dv_installation`
for an existing `DataverseFileInfo` object
2. If there’s no object, start a `DataverseFileInfo` object from the inputs in `DataverseHandoff` and `OpenDPUser`
3. If the fields `dataset_schema_info` and  `file_schema_info`  are not populated, retrieve the JSON-LD information from Dataverse—examples to do this are below.
	- Save the JSON-LD information in `dataset_schema_info`
		- Example of retrieving the JSON-LD info: [lines 125-130](https://github.com/opendp/opendp-ux/blob/develop/server/opendp_apps/dataverses/dataverse_request_handler.py#L116)
	- From the `dataset_schema_info`, pull out the file specific information and save it to `file_schema_info`
		- Example: [lines 132-139](https://github.com/opendp/opendp-ux/blob/develop/server/opendp_apps/dataverses/dataverse_request_handler.py#L116)
4. Return the `DataverseFileInfo`  as JSON. 
	- Exclude the  `data_profile` and `source_file`  fields

### LINKED ISSUE #151: DV User fix
- DataverseUser
  - **create** / POST
     - **input**: `OpenDPUser.object_id`, `DataverseHandoff.object_id`
     - **output**: `DataverseUser.object_id` 
     - **error**:  will return an error if there's an existing `DataverseUser` 
   - **update** / PUT
     - **input**: `OpenDPUser.object_id`, `DataverseHandoff.object_id`
     - **output**: `DataverseUser.object_id` 
     - **error** if `DataverseUser` ​isn't found

- **bug**: Two OpenDPUserSerializer objects!!
  - in user/serializers
  - in dataverses/serializers (delete!!!)

### PR DESCRIPTION
**Pull request by ecowan**

### CODE REVIEW

File: server/opendp_apps/dataverses/views/dataverse_file_view.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+from opendp_apps.dataverses.dataverse_request_handler import DataverseRequestHandler
+from opendp_apps.dataverses.models import DataverseHandoff
+from opendp_apps.dataverses.serializers import DataverseFileInfoSerializer
+from opendp_apps.user.models import DataverseUser
+
+
+class DataverseFileView(viewsets.ViewSet):
+
+    def list(self, request):
+        handoff_id = request.query_params.get('handoff_id')
```
Comment (raprasad) [inline server/opendp_apps/dataverses/views/dataverse_file_view.py:24]: Needs error checking. e.g. Line 17 could blow up on DataverseUser.DoesNotExist, etc.
- example: http://127.0.0.1:8000/api/dv-file/

File: server/opendp_apps/dataverses/views/dataverse_file_view.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        user_id = request.query_params.get('user_id')
+        dataverse_user = DataverseUser.objects.get(object_id=user_id)
+        opendp_user = dataverse_user.user
+        registered_dataverse = dataverse_user.dv_installation
+        # print(handoff_id, user_id)
+        handoff = DataverseHandoff.objects.get(object_id=handoff_id)
```
Comment (raprasad) [inline server/opendp_apps/dataverses/views/dataverse_file_view.py]: Needs error checking; can blow up on DataverseHandoff.DoesNotExist

**Review by raprasad — CHANGES_REQUESTED** [review]

File: server/opendp_apps/dataverses/views/dataverse_user_view.py [hunk 3]
```diff
@@ -28,79 +27,72 @@ def create(self, request):
...
-
-        dataverse_user_serializer = DataverseUserSerializer(data=request.data, context={'request': request})
-        if not dataverse_user_serializer.is_valid():
-            print("INVALID SERIALIZER")
-            return Response(dataverse_user_serializer.errors,
+                             'message': f'OpenDPUser not found with object_id {user_id}'},
                             status=status.HTTP_400_BAD_REQUEST)
 
+        request.data['handoff'] = handoff_id
+        request.data['user'] = user_id
```
Comment (raprasad) [inline server/opendp_apps/dataverses/views/dataverse_user_view.py:33]: The `request.data['user']` attribute is also set on line 36

File: server/opendp_apps/dataverses/views/dataverse_user_view.py [hunk 4]
```diff
@@ -28,79 +27,72 @@ def create(self, request):
...
-                            status=status.HTTP_400_BAD_REQUEST)
-
+        handoff_id = request.data.get('dv_handoff')
         try:
             handoff = DataverseHandoff.objects.get(object_id=request.data.get('dv_handoff'))
         except DataverseHandoff.DoesNotExist:
             return Response({'success': False, 'message': 'No Handoff found'}, status=status.HTTP_400_BAD_REQUEST)
         try:
-            request.data['user'] = OpenDPUser.objects.get(object_id=dataverse_user.user.object_id)
+            request.data['user'] = OpenDPUser.objects.get(object_id=user_id)
```
Comment (raprasad) [inline server/opendp_apps/dataverses/views/dataverse_user_view.py]: Could this be done via the serializer?
Something like:
```python
    def validate_user(self, value):
        """
        Retrieve the OpenDPUser based on the object_id
        """
        try:
            user_obj = OpenDPUser.objects.get(object_id=str(value))
        except OpenDPUser.DoesNotExist:
            raise serializers.ValidationError(f"OpenDP User does not exist for object_id: {value}")

        return value
```

File: server/opendp_apps/dataverses/views/dataverse_user_view.py [hunk 5]
```diff
@@ -28,79 +27,72 @@ def create(self, request):
...
-        try:
-            dataverse_user = DataverseUser.objects.get(object_id=request.data.get('user'))
-        except DataverseUser.DoesNotExist:
-            return Response({'success': False,
-                             'message': f'DataverseUser not found with object_id {user_id}'},
-                            status=status.HTTP_400_BAD_REQUEST)
-
+        handoff_id = request.data.get('dv_handoff')
         try:
             handoff = DataverseHandoff.objects.get(object_id=request.data.get('dv_handoff'))
```
Comment (raprasad) [inline server/opendp_apps/dataverses/views/dataverse_user_view.py]: Could this be moved to the DataverseUserSerializer? e.g. 
```python
    def validate_dv_handoff(self, value):
        """
        Retrieve the DataverseHandoff based on the object_id
        """
        try:
            dv_handoff = DataverseHandoff.objects.get(object_id=str(value))
        except DataverseHandoff.DoesNotExist:
            raise serializers.ValidationError(f"DataverseHandoff does not exist for object_id: {value}")

        return dv_handoff
```

**Review by raprasad — CHANGES_REQUESTED** [review]

File: server/opendp_apps/dataverses/views/dataverse_user_view.py (hunk 5 above)
Comment (ecowan) [reply server/opendp_apps/dataverses/views/dataverse_user_view.py]: I've added a function called `get_object_or_error_response` which essentially does this

**Review by raprasad — APPROVED** [review]
approved

**Pull request merged**